### PR TITLE
Fixes + Channel Selection for CHiME-7 Task 

### PR DESCRIPTION
**Pull request by popcornell**

### CODE REVIEW

File: egs2/chime7_task1/asr1/local/gss_micrank.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+import torchaudio
+import torch
+from copy import deepcopy
+import tqdm
+from pathlib import Path
+import os
+from torch.utils.data import DataLoader, Dataset
+
+
+class EnvelopeVariance(torch.nn.Module):
```
Comment (popcornell) [inline egs2/chime7_task1/asr1/local/gss_micrank.py:12]: Channel selection is based on Envelope variance right now. 
It is not guarantee it will work because of overlapped speech

File: egs2/chime7_task1/asr1/local/install_dependencies.sh [hunk 2]
```diff
@@ -23,10 +23,10 @@ ${MAIN_ROOT}/tools/installers/install_s3prl.sh
 
 if ! command -v gss &>/dev/null; then
   conda install -yc conda-forge cupy=10.2
-  ${MAIN_ROOT}/tools/installers/install_gss.sh.
+  ${MAIN_ROOT}/tools/installers/install_gss.sh
 fi
 
-sox_conda=`command -v ../../../tools/venv/bin/sox 2>/dev/null`
+sox_conda=`command -v $(dirname $(which python))/sox 2>/dev/null`
```
Comment (popcornell) [inline egs2/chime7_task1/asr1/local/install_dependencies.sh:29]: hopefully this fixes the sox issue.

Comment (kamo-naoyuki) [reply egs2/chime7_task1/asr1/local/install_dependencies.sh:29]: FYI. Conda setups some useful shell environment variables: CONDA_PREFIX, CONDA_EXE, etc.

If sox was installed by sox, the path should be ${CONDA_PREFIX}/bin/sox

Comment (popcornell) [reply egs2/chime7_task1/asr1/local/install_dependencies.sh:29]: Many thanks I was not aware of CONDA_PREFIX. Seems much better to use that, it is more clean

Comment (popcornell) [reply egs2/chime7_task1/asr1/local/install_dependencies.sh]: Updated. Also updated [local/generate_chime6_data.sh] with same variable (https://github.com/espnet/espnet/pull/4934/files#diff-83abe1d71c47a59774fbbdc7721c39519004e681d12764a55c81c66f9ffaceae)

**Comment by popcornell** [comment]
```bash
./run.sh --chime6-root YOUR_PATH_TO_CHiME6 --dipco-root PATH_WHERE_DOWNLOAD_DIPCO \
--mixer6-root YOUR_PATH_TO_MIXER6 --stage 0 --ngpu YOUR_NUMBER_OF_GPUs \
--use-pretrained popcornell/chime7_task1_asr1_baseline \
--decode_only 1 
```
This now runs channel selection + GSS + ASR decoding with the pre-trained model.

**Comment by popcornell** [comment]
This is ready for review. 
Automatic Channel Selection now improves further the results (top 80% on each dataset): 

<table>
<thead>
  <tr>
    <th>Dataset</th>
    <th>split</th>
    <th>front-end</th>
    <th>SA-WER (%)</th>
    <th>macro SA-WER (%)</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td>CHiME-6</td>
    <td rowspan="3">dev</td>
    <td>GSS (EV top 80%)</td>
    <td>34.5</td>
    <td rowspan="3">30.8</td>
  </tr>
  <tr>
    <td>DiPCo</td>
    <td>GSS (EV top 80%)</td>
    <td>36.8</td>
  </tr>
  <tr>
    <td>Mixer-6</td>
    <td>GSS (EV top 80%)</td>
    <td>21.2</td>
  </tr>
</tbody>
</table>

vs (classic way) using all outer mics on CHiME-6

<table>
<thead>
  <tr>
    <th>Dataset</th>
    <th>split</th>
    <th>front-end</th>
    <th>SA-WER (%)</th>
    <th>macro SA-WER (%)</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td>CHiME-6</td>
    <td rowspan="3">dev</td>
    <td>GSS (all outer)</td>
    <td>35.5</td>
    <td rowspan="3">31.8</td>
  </tr>
  <tr>
    <td>DiPCo</td>
    <td>GSS (all outer)</td>
    <td>39.3</td>
  </tr>
  <tr>
    <td>Mixer-6</td>
    <td>GSS (all)</td>
    <td>20.6</td>
  </tr>
</tbody>
</table>

File: egs2/chime7_task1/asr1/local/install_dependencies.sh (hunk 2 above)
Comment (popcornell) [reply egs2/chime7_task1/asr1/local/install_dependencies.sh]: @kamo-naoyuki I followed your suggestion and added a script + JSON file to check the MD5 checksum for each file https://github.com/espnet/espnet/blob/cfbb957d9c71c5c7aed27a1d4b2b85b62721381a/egs2/chime7_task1/asr1/local/check_data_gen.py but I had also to add a .json file to this recipe. Is it ok ?

File: egs2/librispeech/ssl1/local/hubert_feature_loader.py [hunk 3]
```diff
@@ -139,12 +139,9 @@ def get_feats(self, data, ref_len=None):
             x = torch.from_numpy(x).float().to(self.device)
             x = x.view(1, -1)
 
-            feat = self.model.wav2vec2.extract_features(
```
Comment (popcornell) [inline egs2/librispeech/ssl1/local/hubert_feature_loader.py:142]: I had to apply black here otherwise tests were failing.

File: espnet/nets/chainer_backend/transformer/mask.py [hunk 4]
```diff
@@ -12,6 +12,8 @@ def make_history_mask(xp, block):
     """
     batch, length = block.shape
     arange = xp.arange(length)
-    history_mask = (arange[None] <= arange[:, None])[None,]
```
Comment (popcornell) [inline espnet/nets/chainer_backend/transformer/mask.py:15]: ditto

**Comment by popcornell** [comment]
Some tests are failing because of some broken dowload.

**Comment by popcornell** [comment]
@sw005320 @simpleoier can we merge this PR ?

**Comment by sw005320** [comment]
I quickly scan the changes, and it looks good to me.
After the CI check, I'll merge it.
I'll try to find time to check more details later.

**Comment by popcornell** [comment]
If somebody can test it, it would be great. 
I did it on my side, but there are lot of stuff going on and some issues could be encountered for sure. 
Thanks to @kamo-naoyuki help and @YoshikiMas for testing.

**Comment by kamo-naoyuki** [comment]
I checked  local/check_data_gen.py, but  I got

```
MD5 Checksum for mixer6/transcriptions/train_call/20090717_104045_LDC_120312.json is not the same. Data has not been generated correctly.You can retry to generate it or re-download it.If this does not work, please reach us. 
MD5 Checksum for mixer6/transcriptions/train_call/20090805_100942_LDC_120221.json is not the same. Data has not been generated correctly.You can retry to generate it or re-download it.If this does not work, please reach us. 
MD5 Checksum for mixer6/transcriptions/train_call/20090810_160537_LDC_120235.json is not the same. Data has not been generated correctly.You can retry to generate it or re-download it.If this does not work, please reach us. 
...
``` 

(Maybe all json files under train_call are failed)  I think some problems still exist in this recipe.

**Comment by kamo-naoyuki** [comment]
For quick check, I'll paste the head of my json file here: mixer6/transcriptions/train_call/20090921_130118_LDC_120501.json

```
[
    {
        "start_time": "1956.580",
        "end_time": "1957.290",
        "words": "hi",
        "speaker": "120501"
    },
    {
        "start_time": "1957.720",
        "end_time": "1959.310",
        "words": "i can\u2019t really hear",
        "speaker": "120501"
    },
    {
        "start_time": "1960.420",
        "end_time": "1962.030",
        "words": "lena",
        "speaker": "120501"
    },
```

I'm suspicious about the `\u2019`. It is correct?

**Comment by popcornell** [comment]
For sure, @kamo-naoyuki can you give me your path to `train_call` relative to your root folder as provided by LDC ? 

\u2019 is correct, if it appears in the original data as provided by LDC.  It should however disappear after text normalization as this is applied https://github.com/espnet/espnet/blob/5d4615f4264574b52c9458c2d1ddc52814400847/egs2/chime7_task1/asr1/local/gen_task1_data.py#L39

Do you still have after running the data generation step ?

**Comment by kamo-naoyuki** [comment]
> For sure, @kamo-naoyuki can you give me your path to train_call relative to your root folder as provided by LDC ?

splits/train_call

**Comment by popcornell** [comment]
I have the same now. But this does not really explain why the MD5 checksum is not the same. 
There must be some difference in the JSON at this point. Can you send me it in Slack, so I can compare ?

**Comment by popcornell** [comment]
UPDATE: the annotation I have differs from one LDC gave to participants (in a non significant way but it changes the hash).
I will recompute the hash with LDC annotation. 
That is what I can do. 
Also the check_data script should better return an error rather than printing.

**Comment by popcornell** [comment]
@kamo-naoyuki Did the check fail for other stuff too ?

File: egs2/chime7_task1/asr1/README.md [hunk 5]
```diff
@@ -158,69 +164,194 @@ chime7_task1
...
 
+## Acknowledgements
 
-
-[sox]:
-[google_group]: 
-[gpu_gss]:
-[gss]: 
-
+We would like to thank Dr. Naoyuki Kamo for his precious help.
```
Comment (kamo-naoyuki) [inline egs2/chime7_task1/asr1/README.md:354]: Thank you for mention me, but sorry, I don't have  Ph.D. :->

Comment (popcornell) [reply egs2/chime7_task1/asr1/README.md:354]: Oh I am sorry. I can remove it. Honestly you deserve an honorary one ;)
